pkg/hubble/exporter/exporter.go

```diff
@@ -0,0 +1,113 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exporter
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	observerAPI "github.com/cilium/cilium/api/v1/observer"
+	"github.com/cilium/cilium/pkg/hubble/exporter/exporteroption"
+
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc/metadata"
+	"gopkg.in/natefinch/lumberjack.v2"
+)
+
+var _ observerAPI.Observer_GetFlowsServer = (*exporter)(nil)
+
+// exporter implements Observer_GetFlowsServer interface. Instead of sending responses over
+// a socket it writes them to a file in JSON format.
+type exporter struct {
+	logger   logrus.FieldLogger
+	observer observerAPI.ObserverServer
+	encoder  *json.Encoder
+}
+
+// NewExporter initializes an exporter.
+func NewExporter(
+	logger logrus.FieldLogger,
+	observer observerAPI.ObserverServer,
+	options ...exporteroption.Option) (*exporter, error) {
+	opts := exporteroption.Default // start with defaults
+	for _, opt := range options {
+		if err := opt(&opts); err != nil {
+			return nil, fmt.Errorf("failed to apply option: %v", err)
+		}
+	}
+	logger.WithField("options", opts).Info("Configuring Hubble event exporter")
+	encoder := json.NewEncoder(&lumberjack.Logger{
+		Filename:   opts.Path,
+		MaxSize:    opts.MaxSizeMB,
+		MaxBackups: opts.MaxBackups,
+		Compress:   opts.Compress,
+	})
+	return newExporter(logger, observer, encoder), nil
+}
+
+func newExporter(logger logrus.FieldLogger, observer observerAPI.ObserverServer, encoder *json.Encoder) *exporter {
+	return &exporter{
+		logger:   logger,
+		observer: observer,
+		encoder:  encoder,
+	}
+}
+
+// Start calls GetFlows and writes responses to a file.
+func (e *exporter) Start() {
+	// Wait for Hubble to observe at least 2 flows to prevent ring.LastWriteParallel()
+	// from underflowing.
+	for {
+		status, err := e.observer.ServerStatus(context.Background(), &observerAPI.ServerStatusRequest{})
+		if err == nil && status.NumFlows > 1 {
+			break
+		}
+		e.logger.WithError(err).Debug("Waiting for Hubble to start...")
+		time.Sleep(1 * time.Second)
+	}
+	req := observerAPI.GetFlowsRequest{Follow: true}
+	if err := e.observer.GetFlows(&req, e); err != nil {
+		e.logger.WithError(err).Warn("Hubble export stopped.")
+	}
+}
+
+func (e *exporter) Send(event *observerAPI.GetFlowsResponse) error {
+	return e.encoder.Encode(event)
+}
+
+func (e *exporter) SetHeader(metadata.MD) error {
+	return nil
+}
+
+func (e *exporter) SendHeader(metadata.MD) error {
+	return nil
+}
+
+func (e *exporter) SetTrailer(metadata.MD) {
+}
+
+func (e *exporter) Context() context.Context {
+	return context.Background()
+}
+
+func (e *exporter) SendMsg(_ interface{}) error {
+	return nil
+}
+
+func (e *exporter) RecvMsg(_ interface{}) error {
+	return nil
+}
```

Comment:
What's all this stubbing?

Reply:
these are stubs for grpc part of `Observer_GetFlowsServer` interface: https://github.com/michi-covalent/cilium/blob/28309a9cfd4cf38dceb86ca58fc20b5e460d112f/pkg/hubble/exporter/exporter.go#L33 i'll add a comment to make it clearer.

Reply:
well if we add another hook they'll go away 😉

Reply:
removed